beluga/include/beluga/algorithm/adaptive_probability_estimator.hpp

```diff
@@ -0,0 +1,100 @@
+// Copyright 2024 Ekumen, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef BELUGA_ALGORITHM_ADAPTIVE_PROBABILITY_ESTIMATOR_HPP
+#define BELUGA_ALGORITHM_ADAPTIVE_PROBABILITY_ESTIMATOR_HPP
+
+#include <beluga/algorithm/exponential_filter.hpp>
+#include <beluga/type_traits/particle_traits.hpp>
+#include <beluga/views/particles.hpp>
+
+#include <range/v3/numeric/accumulate.hpp>
+
+namespace beluga {
+
+/// Random particle probability estimator.
+/**
+ * This class implements an estimator for what probability to use for injecting random
+ * particles (not sampled directly from the particle set) during the resampling step of a
+ * particle filter.
+ *
+ * The addition of random samples allows the filter to recover in case it converged to
+ * the wrong estimate, adding an additional level of robustness.
+ *
+ * This estimator averages the total weight of the particles and computes the ratio
+ * between a short-term and a long-term average over time.
+ *
+ * See Probabilistic Robotics \cite thrun2005probabilistic, Chapter 8.3.3.
+ */
+class AdaptiveProbabilityEstimator {
+ public:
+  /// Constructor.
+  /**
+   * \param alpha_slow Decay rate for the long-term average.
+   * \param alpha_fast Decay rate for the short-term average.
+   */
+  constexpr AdaptiveProbabilityEstimator(double alpha_slow, double alpha_fast) noexcept
+      : slow_filter_{alpha_slow}, fast_filter_{alpha_fast} {
+    assert(0 < alpha_slow);
+    assert(alpha_slow < alpha_fast);
+  }
+
+  /// Reset the internal state of the estimator.
+  /**
+   * It is recommended to reset the estimator after injecting random particles
+   * to avoid spiraling off into complete randomness.
+   */
+  constexpr void reset() noexcept {
+    slow_filter_.reset();
+    fast_filter_.reset();
+    probability_ = 0.0;
+  }
+
+  /// Update the estimation based on a particle range.
+  template <class Range>
+  constexpr void update(Range&& range) {
+    static_assert(ranges::sized_range<Range>);
+    static_assert(beluga::is_particle_range_v<Range>);
+    const auto size = static_cast<double>(range.size());
```

Comment:
@nahueespinosa nit: consider checking for zero size in its integer form, and only casting when computing a weight average. Checking floating point numbers for equality is ill-defined and compilers sometimes emit warnings when you do.

Reply:
Done!